src/library/scala/collection/immutable/List.scala

```diff
@@ -328,28 +332,34 @@ sealed abstract class List[+A] extends AbstractSeq[A]
   }
 
   final override def flatMap[B, That](f: A => GenTraversableOnce[B])(implicit bf: CanBuildFrom[List[A], B, That]): That = {
-    if (bf eq List.ReusableCBF) {
+    if (isLikeListReusableCBF(bf)) {
       if (this eq Nil) Nil.asInstanceOf[That] else {
         var rest = this
-        var found = false
-        var h: ::[B] = null
-        var t: ::[B] = null
+        class Appender extends Function1[B, Unit] {
+          var h: ::[B] = null
+          var t: ::[B] = null
+          override def apply(b: B): Unit = {
+              if (h eq null) {
+                h = new ::(b, Nil)
+                t = h
+              }
+              else {
+                val nx = new ::(b, Nil)
+                t.tl = nx
+                t = nx
+              }
+          }
+        }
+        var appender: Appender = null
         while (rest ne Nil) {
-          f(rest.head).seq.foreach{ b =>
-            if (!found) {
-              h = new ::(b, Nil)
-              t = h
-              found = true
-            }
-            else {
-              val nx = new ::(b, Nil)
-              t.tl = nx
-              t = nx
-            }
+          val c = f(rest.head)
+          if (!c.isEmpty) {
```

Comment:
`GenTraversableOnce.isEmpty` in 2.12.x isn't guaranteed to be cheap. You could add a special case for `eq Nil` as that will be a output type `f`.

Reply:
I don't see that `c` is guaranteed to be a `List`. I have pushed a change to not always call `isEmpty` for some classes but please confirm if these are the right classes

Reply:
reworked